<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
  <head>
    <title>MS-GF+ ChangeLog</title>
    <link type="text/css" rel="stylesheet" href="style.css" />
  </head>

  <body>
    <h1 class="pagetitle">MS-GF+ ChangeLog</h1>
    <div class="pageEditData">
      Added by Sangtae Kim, last edited by Sangtae Kim on Jul 16, 2014
    </div>
    <a href="index.html">MS-GF+ Documentation home</a>
    
    <p>
      <b>06/29/2016 v20160629</b>
    </p>
    <ul>
      <li>Clean up the mzid output a little, reducing file size (v20160615 introduced a change that resulted in larger mzid files)</li>
      <li>Reduced the amount of unnecessary data in the .jar file, cutting size by ~1/2</li>
    </ul>
    
    <p>
      <b>06/15/2016 v20160615</b>
    </p>
    <ul>
      <li>Fix the mzid output when the modification is unknown to unimod. (cvRef now correctly references PSI-MS, and the value will be the name provided in the Mods.txt file)</li>
      <li>Update the unimod.obo to date 2016:02:01, and change code to parse it properly after updating</li>
      <li>Bump some dependencies to a newer version, and change code to properly work with those changes.</li>
    </ul>
    
    <p>
      <b>05/25/2016 v20160525</b>
    </p>
    <ul>
      <li>Output the residue letter and mass of any custom amino acids to the "MassTable" portion of the mzid.</li>
      <li>Update the URIs for the ontologies to their current locations (psi-ms.obo and unit.obo).</li>
    </ul>
    
    <p>
      <b>02/12/2016 v20160212</b>
    </p>
    <ul>
      <li>Ensure that ETciD and EThcD are handled as ETD when using a _dta.txt file with a _ScanType.txt file</li>
    </ul>
    
    <p>
      <b>01/29/2016 v20160129</b>
    </p>
    <ul>
      <li>Added ability to enter custom amino acids using the Mods.txt file.</li>
    </ul>
    
    <p>
      <b>01/21/2016 v20160121</b>
    </p>
    <ul>
      <li>Changed the versioning system - SVN revisions don't work for Git repositories. Now commit date is used for the version.</li>
      <li>Improve the overall progress reporting.</li>
      <li>Cause entire search to fail if a thread exits with a failure.</li>
      <li>Retention Time/Scan Start Time is now output to the mzid file for searches performed on mzML files, as long as the data is available.</li>
    </ul>
    
    <p>
      <b>07/16/2014 v10089</b>
    </p>
    <ul>
      <li>Fixed a bug that crashes when C-term mod mass is below -57Da.</li>
      <li>Added a test that checks whether the output path is valid before processing the data.</li>
    </ul>

    <p>
      <b>06/30/2014 v10072</b>
    </p>
    <ul>
      <li>Optimization for multithreaded performance.</li>
      <li>Several minor bug fixes.</li>
    </ul>

    <p>
      <b>02/10/2014 v9949</b>
    </p>
    <ul>
      <li>New scoring parameters are added for HCD/Q-Exactive/Trypsin/TMT. Parameters for HCD/HighRes/Trypsin/TMT have also been changed. As a result, for HCD spectra of TMT peptides, the number of identifications has been significantly increased.</li>
      <li>MS-GF+ now automatically recognize "TMT6plex" in the modification file and change the protocol to TMT.</li>
      <li>For mzML files converted from Thermo raw data using msconvert, MS-GF+ now reads monoisotopic precursor m/z first from a user parameter (<span class="red">[Thermo Trailer Extra]Monoisotopic M/Z:</span>) instead of the CvParam "MS:1000744".</li>
      <li>The maximum number of variable modifications (per peptide) is written to the mzIdentML output file.</li>
      <li>Previously MS-GF+ ignored spectra having less than <b>20 peaks</b> for non-TOF spectra. Now it ignores spectra having less than <b>10 peaks</b>.</li>
    </ul>

    <p>
      <b>08/28/2013 v9881</b>
    </p>
    <ul>
      <li>Change in database indexing format. The index file keeps non-standard amino acids (characters other than 20 standard residue characters). Previously, non-standard amino acids were converted into ‘?’ while indexing. It caused a problem when converting mzid into pepXML using idconvert (ProteoWizard).</li>
      <li>Fixed a bug that creates duplicate PeptideEvidence items when more than one matches are reported for a spectrum (e.g. -n 10),</li>
      <li>Fixed a bug that crashes MS-GF+ when a modification makes the amino acid mass 0 or smaller (e.g. M-131).</li>
    </ul>

    <p>
      <b>08/28/2013 v9733</b>
    </p>
    <ul>
      <li>Added parameters for CID-LowRes-NoCleavage.</li>
      <li>Misc bug fixes.</li>
    </ul>

    <p>
      <b>04/03/2013 v9517</b>
    </p>
    <ul>
      <li>Previously separate SpectrumIdentificationItems were created for the same peptide if "pre" is different (e.g. R.SIPDSMNYGDEEENK and K.SIPDSMNYGDEEENK). Now they show up as the same SpectrumIdentificationItem. If the score is different due to different NTT (e.g. G.SIPDSMNYGDEEENK), a separate SpectrumIdentificationItem is created.</li>
      <li>Bug fix: falling into an infinite loop while reading an mzML file if it contains &lt;binaryDataArray&gt; of encodedLength="0"</li>
    </ul>

    <p>
      <b>04/03/2013 v9501</b>
    </p>
    <ul>
      <li>Previously, spectra are ignored in the search if the number of peaks is less than 20 for all types spectra. Now, for TOF spectra (i.e. -inst 2), this number has been changed to 3.</li>
    </ul>

    <p>
      <b>04/02/2013 v9494</b>
    </p>
    <ul>
      <li>The following features are added in the SpectrumIdentificationItem when "-addFeatures 1"
        <ul>
          <li>MeanErrorAll: Mean of mass errors of explained fragment ion peaks</li>
          <li>StdevErrorAll: Standard deviation of mass errors of explained fragment ion peaks</li>
          <li>MeanErrorTop7: Mean of mass errors of 7 fragment ion peaks with highest intensities</li>
          <li>StdevErrorTop7: Standard deviation of mass errors of 7 fragment ion peaks with highest intensities</li>
        </ul>
      </li>
    </ul>

    <p>
      <b>03/25/2013 v9436</b>
    </p>
    <ul>
      <li>Added TMT scoring model for HCD/HighRes (-m 3 -inst 1)</li>
      <li>Bug fix: duplicate PeptideEvidence id in mzid output</li>
      <li>Bug fix: crashing while using user-trained scoring parameters</li>
    </ul>

    <p>
      <b>03/05/2013 v9324</b>
    </p>
    <ul>
      <li>Added Q-Exactive unlabeled phosphorylation parameters</li>
      <li>Bug fix: recognizing ETD+SA spectra in mzML as CID</li>
    </ul>

    <p>
      <b>02/27/2013 v9324</b>
    </p>
    <ul>
      <li>Minor bug fix: mistakenly assigning 3 mods to a N-term amino acid.</li>
    </ul>

    <p>
      <b>02/15/2013 v9312</b>
    </p>
    <ul>
      <li>Added scoring parameter sets for Q-Exactive iTRAQ (-inst 3 -protocol 2) and iTRAQ phosphopeptide enriched (-inst 3 -protocol 3) samples.</li>
    </ul>

    <p>
      <b>02/15/2013 v9284</b>
    </p>
    <ul>
      <li>When "-protocol" parameter is missing, MS-GF+ automatically selects an appropriate protocol depending on the modification file.</li>
      <li>In the mzid output, Enzyme has attributes "missedCleavages".</li>
    </ul>

    <p>
      <b>02/14/2013 v9249</b>
    </p>
    <ul>
      <li>Scoring parameters for Q-Exactive (-inst 3) have been added.</li>
    </ul>

    <p>
      <b>02/04/2013 v9244</b>
    </p>
    <ul>
      <li>A bug (crash with an exception) in edu.ucsd.msjava.ui.ScoringParamGen has been fixed.</li>
    </ul>

    <p>
      <b>01/03/2013 v9176</b>
    </p>
    <ul>
      <li>"-ti" parameter accepts only two comma separated integers, i.e., "-ti 1" will be rejected.</li>
      <li>MzIDToTsv adds "Title" column to the tsv file if the input spectrum is the mgf format.</li>
    </ul>

    <p>
      <b>12/19/2012 v9107</b>
    </p>
    <ul>
      <li>"No enzyme" (-e 1) was renamed to "unspecific cleavage".</li>
      <li>"No enzyme (peptidomics)" (-e 9) was renamed to "no cleavage".</li>
      <li>&lt;Enzyme&gt;...&lt;/Enzyme&gt; in the mzid output now shows right CVs for -e 1 and -e 9.</li>
      <li>The following bug has been fixed
        <ul>
          <li>Precursor errors are negated when MzIDToTsv was run.</li>
          <li>C-term specific modifications do not show up in result files.</li>
        </ul>
      </li>
    </ul>

    <p>
      <b>12/10/2012 v9014</b>
    </p>
    <ul>
      <li>The following bug has been fixed
        <ul>
          <li>Ignoring some spectra when -n value is larger than 1</li>
        </ul>
      </li>
    </ul>

    <p>
      <b>11/30/2012 v9012</b>
    </p>
    <ul>
      <li>The following bugs have been fixed
        <ul>
          <li>Profile spectra should be ignored but not.</li>
          <li>Peptides shorter than MinLength are reported.</li>
          <li>Redundant proteins are reported in the same line in tsv files</li>
          <li>Some SpectrumIdentificationResult elements have no SpectrumIdentificationItem element.</li>
          <li>Different PeptideEvident elements have the same ID.</li>
        </ul>
      </li>
    </ul>

    <p>
      <b>11/09/2012 v8884</b>
    </p>
    <ul>
      <li>Bug fix: crashing while reading mzML files converted from wiff</li>
    </ul>

    <p>
      <b>11/09/2012 v8873</b>
    </p>
    <ul>
      <li>Fixed a bug of ignoring N-term peptide when N-term Met cleaved peptide exists.</li>
      <li>Add scan numbers for NIST mgf files</li>
    </ul>

    <p>
      <b>10/30/2012 v8806</b>
    </p>
    <ul>
      <li>Fixed a bug in edu.ucsd.msjava.misc.MS2ToMgf</li>
    </ul>

    <p>
      <b>10/29/2012 v8792</b>
    </p>
    <ul>
      <li>Fixed bugs to ignore N-term fixed mods in the output</li>
      <li>Added edu.ucsd.msjava.misc.MS2ToMgf</li>
    </ul>

    <p>
      <b>10/11/2012 v8719</b>
    </p>
    <ul>
      <li>Fixed bugs reporting NaN as additional features when spectrum has charge 0</li>
    </ul>

    <p>
      <b>10/04/2012 v8605</b>
    </p>
    <ul>
      <li>Updates to handle multiple charge states in the ms2 format</li>
    </ul>

    <p>
      <b>10/03/2012 v8597</b>
    </p>
    <ul>
      <li>Fixed the bug to erroneously report precursorMz in converted tsv files</li>
    </ul>

    <p>
      <b>09/26/2012 v8540</b>
    </p>
    <ul>
      <li>Update MS2 parser to support <a href="http://noble.gs.washington.edu/proj/crux/ms2-format.html">http://noble.gs.washington.edu/proj/crux/ms2-format.html</a>
      </li>
      <li>Old MS2 format (<a href="http://fields.scripps.edu/sequest/SQTFormat.html">http://fields.scripps.edu/sequest/SQTFormat.html</a>) is no longer supported</li>
    </ul>

    <p>
      <b>09/20/2012 v8490</b>
    </p>
    <ul>
      <li>Fix the following bugs:
        <ul>
          <li>Same PeptideEvidence occurs more than once</li>
          <li>SearchModification/SpecificityRules is always empty</li>
          <li>Unknown modification cvParam does not have cvRef</li>
          <li>SpectrumIdentificationResult has no SpectrumIdentificationItem</li>
        </ul>
      </li>
      <li>MzIDToTsv -o TSVFile is now an optional parameter.</li>
    </ul>

    <p>
      <b>09/18/2012 v8477</b>
    </p>
    <ul>
      <li>The extension of target/decoy concatenated database file has changed from .revConcat.fasta to .revCat.fasta.</li>
      <li>*.revConcat.fasta: decoy proteins have prefix "REV_" (used by MS-GFDB)</li>
      <li>*.revCat.fasta: decoy proteins have prefix "XXX_" (used by MS-GF+)</li>
    </ul>

    <p>
      <b>09/18/2012 v8472</b>
    </p>
    <ul>
      <li>Bug fix: N-term or C-term residue-specific modifications (e.g. pyro-glu from Q) will have locations 1 (N-term) or length (C-term).</li>
      <li>Fix a bug to write SpectrumIdentificationResult twice for some spectra without precursor charges</li>
    </ul>

    <p>
      <b>09/17/2012 v8449</b>
    </p>
    <ul>
      <li>Fix a bug in parsing Agilent mzML files</li>
    </ul>

    <p>
      <b>09/13/2012 v8442</b>
    </p>
    <ul>
      <li>Fix the bug ignoring PSMs with IsotopeError=0, when MinIsotopeError was negative</li>
      <li>Change the '-ntt' parameter to set the minimum number of termini following the enzyme specificity rule</li>
      <li>Change the default value of '-ntt' from 1 to 2</li>
      <li>Fix the bug to set "fixedMod=true" for variable modifications</li>
      <li>Fix the bug not showing fixed modifications in the peptide list</li>
      <li>Fix the bug to show the location of N-term modification as 1 instead of 0</li>
      <li>Fix the bug to calculate massToChargeRatio as neutral masses instead of charged masses</li>
      <li>Replace the cvParam showing the dissociation method with a userParam named AssumedDissociationMethod in the SpectrumIdentificationItem</li>
    </ul>

    <p>
      <b>08/30/2012 v8299</b>
    </p>

    <ul>
      <li>Provide MzID to TSV converter (edu.ucsd.msjava.ui.MzIDToTsv)</li>
      <li>Ignore profile spectra in the analysis</li>
      <li>Support JRE 1.6</li>
    </ul>

    <p>
      <b>08/29/2012 v8297</b>
    </p>
    <ul>
      <li>Fix minor bugs</li>
      <li>Add the scan number CV in the output</li>
    </ul>

    <p>
      <b>08/27/2012 v8283</b>
    </p>
    <ul>
      <li>Fix minor bugs</li>
    </ul>
  </body>
</html>
