<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
  <head>
    <title>MS-GF+</title>
    <link type="text/css" rel="stylesheet" href="style.css" />
  </head>

  <body>
    <h1 class="pagetitle">MS-GF+</h1>
    <div class="pageEditData">
      Added by Sangtae Kim, last edited by Sangtae Kim on May 17, 2013
    </div>
    <a href="index.html">MS-GF+ Documentation home</a>
    <h1>MS-GF+</h1>

    <p>
      <font color="#800000">
        <a href="MS-GFDB.html" title="MS-GFDB">(How to migrate from MS-GFDB to MS-GF)</a>
      </font>
    </p>

    <p>
      <font color="#800000">
        <a href="Changelog.html" title="MS-GF+ ChangeLog">ChangeLog</a>
      </font>
    </p>

    <div class="codePanel">
      <pre class="code">Usage: java -Xmx3500M -jar MSGFPlus.jar
	<span class="code-keyword">-s SpectrumFile</span> (*.mzML, *.mzXML, *.mgf, *.ms2, *.pkl or *_dta.txt)
	   Spectra should be centroided. Profile spectra will be ignored.
	<span class="code-keyword">-d DatabaseFile</span> (*.fasta or *.fa)
	<span class="code-keyword">[-o OutputFile (*.mzid)]</span> (<span class="code-object">Default: SpectrumFileName.mzid</span>)
	<span class="code-keyword">[-t PrecursorMassTolerance]</span> (e.g. 2.5Da, 20ppm or 0.5Da,2.5Da, <span class="code-object">Default: 20ppm</span>)
	   Use comma to set asymmetric values. E.g. <span class="code-quote">"-t 0.5Da,2.5Da"</span> will set 0.5Da to the minus (expMass&lt;theoMass) and 2.5Da to plus (expMass&gt;theoMass)
	<span class="code-keyword">[-ti IsotopeErrorRange]</span> (Range of allowed isotope peak errors, <span class="code-object">Default:0,1</span>)
	   Takes into account of the error introduced by chooosing a non-monoisotopic peak for fragmentation.
	   On Windows, put the range inside <span class="code-quote">""</span> (e.g. <span class="code-quote">"0,1"</span>).
	   The combination of -t and -ti determins the precursor mass tolerance.
	   E.g. <span class="code-quote">"-t 20ppm -ti -1,2"</span> tests abs(exp-calc-n*1.00335Da)&lt;20ppm for n=-1, 0, 1, 2.
	<span class="code-keyword">[-thread NumThreads]</span> (Number of concurrent threads to be executed, <span class="code-object">Default: Number of available cores</span>)
	<span class="code-keyword">[-tda 0/1]</span> (<span class="code-object">0: don't search decoy database (Default)</span>, 1: search decoy database)
	<span class="code-keyword">[-m FragmentMethodID]</span> (<span class="code-object">0: As written in the spectrum or CID if no info (Default)</span>, 1: CID, 2: ETD, 3: HCD)
	<span class="code-keyword">[-inst InstrumentID]</span> (<span class="code-object">0: Low-res LCQ/LTQ (Default)</span>, 1: High-res LTQ, 2: TOF, 3: Q-Exactive)
	<span class="code-keyword">[-e EnzymeID]</span> (0: unspecific cleavage, <span class="code-object">1: Trypsin (Default)</span>, 2: Chymotrypsin, 3: Lys-C, 4: Lys-N, 5: glutamyl endopeptidase, 6: Arg-C, 7: Asp-N, 8: alphaLP, 9: no cleavage)
	<span class="code-keyword">[-protocol ProtocolID]</span> (<span class="code-object">0: NoProtocol (Default)</span>, 1: Phosphorylation, 2: iTRAQ, 3: iTRAQPhospho)
	<span class="code-keyword">[-ntt 0/1/2]</span> (Number of Tolerable Termini, <span class="code-object">Default: 2</span>)
	   E.g. For trypsin, 0: non-tryptic, 1: semi-tryptic, 2: fully-tryptic peptides only.
	<span class="code-keyword">[-mod ModificationFileName]</span> (Modification file, <span class="code-object">Default: standard amino acids with fixed C+57</span>)
	<span class="code-keyword">[-minLength MinPepLength]</span> (Minimum peptide length to consider, <span class="code-object">Default: 6</span>)
	<span class="code-keyword">[-maxLength MaxPepLength]</span> (Maximum peptide length to consider, <span class="code-object">Default: 40</span>)
	<span class="code-keyword">[-minCharge MinCharge]</span> (Minimum precursor charge to consider if charges are not specified in the spectrum file, <span class="code-object">Default: 2</span>)
	<span class="code-keyword">[-maxCharge MaxCharge]</span> (Maximum precursor charge to consider if charges are not specified in the spectrum file, <span class="code-object">Default: 3</span>)
	<span class="code-keyword">[-n NumMatchesPerSpec]</span> (Number of matches per spectrum to be reported, <span class="code-object">Default: 1</span>)
	<span class="code-keyword">[-addFeatures 0/1]</span> (<span class="code-object">0: output basic scores only (Default)</span>, 1: output additional features)
Example (high-precision): java -Xmx3500M -jar MSGFPlus.jar -s test.mzXML -d IPI_human_3.79.fasta -t 20ppm -ti "-1,2" -ntt 0 -tda 1 -o testMSGFPlus.mzid
Example (low-precision): java -Xmx3500M -jar MSGFPlus.jar -s test.mzXML -d IPI_human_3.79.fasta -t 0.5Da,2.5Da -ntt 0 -tda 1 -o testMSGFPlus.mzid</pre>
    </div>

    <h5>Parameters:</h5>
    <ul>
      <li>
        <b>-s SpectrumFile</b> (.mzML*, *.mzXML, *.mgf, *.ms2, *.pkl or *_dta.txt) - Required
        <ul>
          <li>Spectrum file name. Currently, MS-GF+ supports the following file formats: mzML, mzXML, mzML, mgf, ms2, pkl and _dta.txt.</li>
          <li>
            <font color="#0000ff">We recommend to use mzML, whenever possible.</font>
          </li>
        </ul>
      </li>
      <br />
      <li>
        <b>-d DatabaseFile</b> (*.fasta or *.fa) - Required
        <ul>
          <li>Path to the protein database file. If the database file does not have auxiliary index files (*.canno, *.cnlcp, *.csarr, and *.cseq), MS-GF+ will create them.</li>
          <li>When "-tda 1" option is used, the database specified here must contain only target protein sequences.</li>
        </ul>
      </li>
    </ul>
    <p class="note">If multiple MS-GF+ processes access the same database file, it is strongly recommended to index the database prior to the database search by running BuildSA (see below).</p>

    <ul>
      <li>
        <b>-o OutputFile</b> (*.mzid)
        <ul>
          <li>Filename where the output (mzIdentML 1.1 format) will be written.</li>
          <li>File extension must be "mzid" (case sensitive).</li>
          <li>By default, the output file name will be "[SpectrumFileName].mzid".</li>
          <li>E.g. for the input spectrum file "test.mzML", the output will be written to "test.mzid" if this parameter is not specified.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-t ParentMassTolerance</b>&nbsp;(Default: 20ppm)
        <ul>
          <li>Parent mass tolerance in Da. or ppm. There must be no space between the number and the unit. E.g. 2.5Da, 20ppm</li>
          <li>To set asymmetric tolerances, use comma to separate left (experimental mass &lt; theoretical mass) or right (experimental mass &gt; theoretical mass) tolerances. E.g. 0.5Da,2.5Da</li>
          <li>It is recommended to use a tight tolerance rather than a loose tolerance (e.g. for Orbitrap data, 10 or 20ppm usually identifies more spectra than 50ppm).</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-ti IsotopeErrorRange</b>&nbsp;(Default: 0,1)
        <ul>
          <li>Takes into account of the error introduced by choosing non-monoisotopic peak for fragmentation.</li>
          <li>If the parent mass tolerance is equal to or larger than 0.5Da or 500ppm, this parameter will be ignored.</li>
          <li>The combination of -t and -ti determins the precursor mass tolerance.</li>
          <li>E.g. "-t 20ppm -ti -1,2" tests abs(exp-calc-n*1.00335Da)&lt;20ppm for n=-1, 0, 1, 2.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-thread NumOfThreads</b> (Number of concurrent threads to be executed, Default: Number of available cores)
        <ul>
          <li>Number of concurrent threads to be executed together.</li>
          <li>Default value is the number of available logical cores (e.g. 8 for quad-core processor with hyper-threading support).</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-tda 0/1</b> (0: don't search decoy database (default), 1: search decoy database to compute FDR)
        <ul>
          <li>Indicates whether to search the decoy database or not.</li>
          <li>If 0, the decoy database is not searched.</li>
          <li>If 1, FDRs are computed based on the target-decoy approach (i.e. reversed database is appended to the target database and MS-GF+ searches the combined database)
            <ul>
              <li>FDR(t) = #(DecoyPSMs with score equal or above t) / #(TargetPSMs with score equal or above t).</li>
              <li>PSM: Peptide-Spectrum Match</li>
              <li>-log(SpecProb) is used as the score to compute FDR.</li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
    <p class="note">If -tda 1 is specified, MS-GF+ automatically creates a combined target/reversed database file (DBFileName.revConcat.fasta). Thus, when specifying "-d" parameter, DatabaseFile must contain only target proteins.</p>

    <ul>
      <li>
        <b>-m FragmentationMethodID</b> (0: as written in the spectrum or CID if no info (Default), 1: CID, 2: ETD, 3: HCD, 4: Merge spectra from the same precursor)
        <ul>
          <li>Fragmentation method identifier (used to determine the scoring model).</li>
          <li>If the identifier is 0 and fragmentation method is written in the spectrum file (e.g. mzML files), MS-GF+ will recognize the fragmentation method and use a relevant scoring model.</li>
          <li>If the identifier is 0 and there is no fragmentation method information in the spectrum (e.g. mgf files), CID model will be used by default.</li>
          <li>If the identifier is non-zero and the spectrum has fragmentation method information, only the spectra that match with the identifier will be processed.</li>
          <li>If the identifier is non-zero and the spectrum has no fragmentation method information, MS-GF+ will process all spectra assuming the specified fragmentation method.</li>
          <li>If the identifier is 4, MS/MS spectra from the same precursor ion (e.g. CID/ETD pairs, CID/HCD/ETD triplets) will be merged and the "merged" spectrum will be used for searching instead of individual spectra. See Kim et al., MCP 2010 for details.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-inst InstrumentID</b> (0: Low-res LCQ/LTQ (Default for CID and ETD), 1: High-res LTQ (Default for HCD), 2: TOF, 3: Q-Exactive)
        <ul>
          <li>Identifier of the instrument to generate MS/MS spectra (used to determine the scoring model).</li>
          <li>For "hybrid" spectra with high-precision MS1 and low-precision MS2, use 0.</li>
          <li>For usual low-precision instruments (e.g. Thermo LTQ), use 0.</li>
          <li>If MS/MS fragment ion peaks are of high-precision (e.g. tolerance = 10ppm), use 2.</li>
          <li>For TOF instruments, use 2.</li>
          <li>For Q-Exactive HCD spectra, use 3.</li>
          <li>For other HCD spectra, use 1.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-e EnzymeID</b> (Default: 1)
        <ul>
          <li>Enzyme identifier. Trypsin (1) will be used by default.</li>
          <li>0: unspecific cleavage, 1: Trypsin (default), 2: Chymotrypsin, 3: Lys-C, 4: Lys-N, 5: glutamyl endopeptidase (Glu-C), 6: Arg-C, 7: Asp-N, 8: alphaLP, 9: no cleavage</li>
          <li>Use 9 for peptidomics studies</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-p ProtocolID</b> (Default: 0)
        <ul>
          <li>Protocol identifier. Protocols are used to enable scoring parameters for enriched and/or labeled samples.</li>
          <li>0: No protocol (Default)</li>
          <li>1: Phosphorylation: for phosphopeptide enriched samples</li>
          <li>2: iTRAQ: for iTRAQ-labeled samples</li>
          <li>3: iTRAQPhospho: for phosphopeptide enriched and iTRAQ-labeled samples</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-ntt 0/1/2</b> (Number of tolerable (tryptic) termini, Default: 2)
        <ul>
          <li>This parameter is used to apply the enzyme cleavage specificity rule when searching the database.</li>
          <li>Specifies the minimum number of termini matching the enzyme specificity rule.
            <ul>
              <li>For example, for trypsin, K.ACDEFGHR.C (NTT=2), G.ACDEFGHR.C (NTT=1), K.ACDEFGHI.C (NTT=1) and G.ACDEFGHR.C (NTT=0).</li>
              <li>'-ntt 2' will search for fully tryptic peptides only.</li>
            </ul>
          </li>
          <li>By default, -ntt 2 will be used. Using -ntt 1 (or 0) will make the search significantly slower.</li>
        </ul>
      </li>
    </ul>
    <p class="note">The meaning and the default value have changed as of version 8442 (Sept. 2012).</p>

    <ul>
      <li>
        <b>-mod ModificationFile</b> (Default: standard amino acids with fixed C+57)]
        <ul>
          <li>Modification file name. ModificationFile contains the modifications to be considered in the search.</li>
          <li>If -mod option is not specified, standard amino acids with fixed Carboamidomethylation C will be used.</li>
          <li>
            <a href="examples/Mods.txt">Download</a> an example modification file.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-minLength MinPepLength</b> (Default: 6)
        <ul>
          <li>Minimum length of the peptide to be considered.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-maxLength MaxPepLength</b> (Default: 40)
        <ul>
          <li>Maximum length of the peptide to be considered.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-minCharge MinPrecursorCharge</b> (Default: 2)
        <ul>
          <li>Minimum precursor charge to consider. This parameter is used only for spectra with no charge.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-maxCharge MinPrecursorCharge</b> (Default: 3)
        <ul>
          <li>Maximum precursor charge to consider. This parameter is used only for spectra with no charge.</li>
        </ul>
      </li>
      <br />
      <li>
        <b>-n NumMatchesPerSpec</b> (Default: 1)
        <ul>
          <li>Number of peptide matches per spectrum to report.</li>
          <li>Expected false discovery rates (EFDRs) will be reported only when this value is 1.</li>
        </ul>
      </li>
      <li>
        <b>-addFeatures 0/1</b>
        <ul>
          <li>If 0, only basic scores are reported.</li>
          <li>If 1, the following features are reported
            <ul>
              <li>MS2IonCurrent: Summed intensity of all product ions</li>
              <li>ExplainedIonCurrentRatio: Summed intensity of all matched product ions (e.g. b, b-H2O, y, etc.) divided by MS2IonCurrent</li>
              <li>NTermIonCurrentRatio: Summed intensity of all matched prefix ions (e.g. b, b-H2O, etc.) divided by MS2IonCurrent</li>
              <li>CTermIonCurrentRatio: Summed intensity of all matched suffix ions (e.g. y, y-H2O, etc.) divided by MS2IonCurrent</li>
            </ul>
          </li>
        </ul>
      </li>
      <li>
        <b>-showQValue 0/1</b>
        <ul>
          <li>If 0, QValue and PepQValue are not reported.</li>
          <li>If 1, QValue (PSM-level Q-value) and PepQValue (peptide-level Q-value) are reported (Default).</li>
          <li>This parameter is ignored when "-tda 0".</li>
        </ul>
      </li>
    </ul>

    <h5>MS-GF+ output</h5>
    <p>MS-GF+ outputs results as an mzIdentML (version 1.1) file. See <a href="http://www.psidev.info/mzidentml/">http://www.psidev.info/mzidentml/</a> for details on the mzIdentML format. For every PSM, MS-GF+ reports the scores.&nbsp;</p>

    <ul>
      <li>
        <b>MS-GF:RawScore</b>: MS-GF+ raw score of the peptide-spectrum match</li>
      <li>
        <b>MS-GF:DeNovoScore</b><b>:</b> the score of the optimal scoring peptide for the spectrum (not necessary in the database)&nbsp;(MS-GF:RawScore &lt;= MS-GF:DeNovoScore)</li>
      <li>
        <b>MS-GF:SpecEValue</b>: spectral E-value (spectrum level E-value) of the peptide-spectrum match - the lower the better</li>
      <li>
        <b>MS-GF:EValue</b>: database level E-value (expected number of peptides in a random database having equal or better scores than the PSM score) - the lower the better</li>
      <li>
        <b>MS-GF:QValue</b>
        <ul>
          <li>PSM-level Q-value estimated using the target-decoy approach.</li>
          <li>MS-GF:QValue is computed solely based on MS-GF:SpecEValue.</li>
        </ul>
      </li>
      <li>
        <b>MS-GF:PepQValue</b>
        <ul>
          <li>Peptide-level Q-value estimated using the target-decoy approach.</li>
          <li>Reported only if "-tda 1" is specified.</li>
          <li>If multiple spectra are matched to the same peptide, only the best scoring PSM (lowest SpecProb) is retained. After that, MS-GF:PepQValue is calculated as #DecoyPSMs&gt;s / #TargetPSMs&gt;s among the retained PSMs. This approximates the Q-value of the set of unique peptides. In the MS-GF+ output, the same PepQValue value is given to all PSMs sharing the peptide. So, even a low-quality PSM may get a low PepQValue (if it has a high-quality "sibling" PSM sharing the peptide). Note that this should not be used to count the number of identified PSMs.</li>
        </ul>
      </li>
      <li>
        <font color="#3366ff">Using MzIDToTsv one can convert MS-GF+ output (*.mzid) into the tsv format</font>
      </li>
    </ul>

    <h5>MS-GF+ output example</h5>
    <p>
      <a href="examples/test.mzid">MzIdentML format</a>
    </p>
    <p>
      <a href="examples/test_Unrolled.tsv">TSV format</a> (converted by MzIDToTsv using MzIDToTsv)
    </p>
  </body>
</html>
